Sources/PubNub/EventEngine/Presence/Presence.swift

```diff
@@ -0,0 +1,158 @@
+//
+//  Presence.swift
+//
+//  PubNub Real-time Cloud-Hosted Push API and Push Notification Client Frameworks
+//  Copyright © 2023 PubNub Inc.
+//  https://www.pubnub.com/
+//  https://www.pubnub.com/terms
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+// MARK: - PresenceState
+
+protocol PresenceState: Equatable {
+  var input: PresenceInput { get }
+}
+
+extension PresenceState {
+  var channels: [String] {
+    input.channels
+  }
+  var groups: [String] {
+    input.groups
+  }
+  
+  func isEqual(to otherState: some PresenceState) -> Bool {
+    (otherState as? Self) == self
+  }
+}
+
+//
+// A namespace for Events, concrete State types and Invocations used in Subscribe EE
+//
+enum Presence {}
+
+// MARK: - Presence States
+
+extension Presence {
+  struct Heartbeating: PresenceState {
+    let input: PresenceInput
+  }
+  
+  struct HeartbeatCooldown: PresenceState {
+    let input: PresenceInput
+  }
+
+  struct HeartbeatReconnecting: PresenceState {
+    let input: PresenceInput
+    let currentAttempt: Int
+    let error: PubNubError
+  }
+  
+  struct HeartbeatFailed: PresenceState {
+    let input: PresenceInput
+    let error: PubNubError
+  }
+  
+  struct HeartbeatStopped: PresenceState {
+    let input: PresenceInput
+  }
+  
+  struct HeartbeatInactive: PresenceState {
+    let input: PresenceInput = PresenceInput()
```

Comment:
@jguz-pubnub In `inactive` state according to the graph we don't have any `input`.

Reply:
I've missed that `input` is part of `PresenceState` protocol conformance. 

In Rust, each state specifies what data it needs and no “general” requirement because some states shouldn't have channels / groups at all and some has additional data.

So it is fine as it now ;)